### PR TITLE
Use tensorized evaluation in MappingQGeneric 

### LINKED ISSUE #4598: Use tensorized evaluation for main work in MappingQGeneric
The functions `maybe_update_Jacobians`, `maybe_compute_q_points`, `maybe_update_jacobian_grads` of `MappingQGeneric` can be made substantially faster for high order if they do not use the naive summation over both quadrature points and the mapping support points at complexity n^{2d} but rather tensorized evaluation at complexity n^{d+1}, for the case that the quadrature is derived from a tensor product. Since our quadrature formulas do not provide this information yet, we need to add the functionality there first. With this ingredient and the functionality in `include/deal.II/matrix_free/evaluation_kernels.h`, this should be not more than a few hundreds lines of code. I hope to get around implementing this during the next months. If someone would like to contribute, I am more than happy to discuss the relevant parts.

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #4598. Based on #4668.

In case the `Quadrature` object that is used to construct a `MappingQGeneric`object is in fact a tensor product of (identical) 1D quadrature formulas, `MappingQGenerics' ` member functions `maybe_update_Jacobians`, `maybe_compute_q_points` and `maybe_update_jacobian_grads` use tensorized evaluation based on the classes in the `internal::MatrixFree` namespace.

For this to pass the entire test suite some test outputs had to be modfied. Currently, only `lac/tensor_product_matrix_03` and ` numerics/project_q_hierarchical_03.debug`are failing for me.

For just calling `fe_values.reinit()` on a triangulation that has been refined globally five times using `MappingQGeneric<3>(6)` I observe a speedup factor of around 10 when computing `quadrature_points`, `jacobians` and `jacobian_grads`.

(edit: extract the changes for `Quadrature` to #4668)

### CODE REVIEW

**Comment by kronbichler** [comment]
Regarding the `tensor_product_matrix_03` see #4665.

**Comment by masterleinad** [comment]
This currently uses the suboptimal `internal::FEEvaluationImpl<internal::MatrixFreeFunctions::tensor_general, dim, -1, 0, n_comp, double>::evaluate`. In a subsequent PR I will replace this by a run time choice of the correct template specialization/instantiation.

File: source/fe/mapping_q_generic.cc [hunk 1]
```diff
@@ -1527,20 +1612,71 @@ namespace internal
...
+                for (unsigned int out_comp=0; out_comp<n_comp-1; ++out_comp)
+                  for (unsigned int i=0; i<n_q_points; ++i)
+                    for (unsigned int in_comp=0; in_comp<vec_length; ++in_comp)
+                      quadrature_points[i][out_comp*vec_length+in_comp]
+                        = data.values_quad[out_comp*n_q_points+i][in_comp];
+                // Treat the last component special as it might not be full.
+                for (unsigned int i=0; i<n_q_points; ++i)
+                  for (unsigned int in_comp=0; in_comp<spacedim-(n_comp-1)*vec_length; ++in_comp)
+                    quadrature_points[i][(n_comp-1)*vec_length+in_comp]
+                      = data.values_quad[(n_comp-1)*n_q_points+i][in_comp];
```
Comment (kronbichler) [inline source/fe/mapping_q_generic.cc]: I don't understand the point of using a separate loop here. All the bounds should be known at compile time, so checking for `in_comp<vec_length && in_comp<spacedim-(n_comp-1)*vec_length` should work, shouldn't it?

Comment (masterleinad) [inline source/fe/mapping_q_generic.cc]: Yes, sure that's a possibilty.

**Comment by kronbichler** [comment]
You may want to squash the last two commits into the first one, though.

**Comment by masterleinad** [comment]
@kronbichler This is ready from my side.

**Comment by kronbichler** [comment]
Very nice! Thanks for the work you put into this. Let us merge this and see how it performs.

**Comment by bangerth** [comment]
I haven't tried to understand the patch in detail, but I do see a lot of new `mutable` variables. Are they all protected by a mutex?

**Comment by kronbichler** [comment]
@bangerth The mutable variable are inside the `InternalData` of `MappingQGeneric` that is held by `FEValues`. Since we require `FEValues` to be thread local, nothing bad can happen. We could argue that we should not mess in the internal data of mapping (as opposed to the output data where we store the result of the transformation) - in that case we should add some scratch array in the output data.

**Comment by bangerth** [comment]
I see -- the `InternalData` objects are indeed not prone to threading problems. But then I don't understand why you make these variables `mutable`. From where do you need to change them where the object you're changing is currently `const`?

**Comment by masterleinad** [comment]
All of them are only used in the `maybe_update_Jacobians`, `maybe_compute_q_points` and `maybe_update_jacobian_grads` and these are called in `MappingQGeneric<dim,spacedim>::fill_fe_values` which is const.

**Comment by bangerth** [comment]
I see, and now remember that we already have this `mutable` pattern for other variables in this context. Thanks for confirming!

**Pull request merged**